packages/react/src/createReactQueryHooks.tsx

```diff
@@ -229,6 +246,28 @@ export function createReactQueryHooks<
     return React.useContext(Context);
   }
 
+  /**
+   * Ugly hack to make sure errors return `status`='error` when doing SSR
+   * @link https://github.com/trpc/trpc/pull/1645
+   */
+  function useSSRError<TQuery extends UseQueryResult | UseInfiniteQueryResult>(
+    queryKey: unknown[],
+    query: TQuery,
+  ): TQuery {
+    const { queryClient, ssrEnabled } = useContext();
+    const isMounted = useIsMounted();
+
+    if (
+      isMounted ||
+      !query.error ||
+      !ssrEnabled ||
+      !queryClient.getQueryCache().find(queryKey)?.meta?._trpcSSR
```

Comment:
Any reason why `meta` isn't available as `useQuery().meta`, @TkDodo?

Reply:
meta is part of the options, not part of the result. Same as the queryKey for example.

Reply:
Could it be both so I don't have to reach into the query cache to find it?

Reply:
> just for my understanding: why are queries not in `error` state when they have an error ?


How we do the SSR prepass to enable SSR:

- Run through the app tree until there are no pending queries: https://github.com/trpc/trpc/blob/f27680fe9ccc444da6282f3e118ba82c3b0b2095/packages/next/src/withTRPC.tsx#L173-L192
- In each query hook, we call `prefetchQuery` when doing the prepass in order to fetch it on the server: https://github.com/trpc/trpc/blob/f27680fe9ccc444da6282f3e118ba82c3b0b2095/packages/react/src/createReactQueryHooks.tsx#L243-L251


Since the `useQuery()` is mounted with some (in this case, erroneous) data in the cache, it will go through the `optimisticResults`-flow which never really sets `status='error'` [as far as I can tell](https://github.com/tannerlinsley/react-query/blob/16783feed17a9361e4263edc8fecf564bc127de7/src/core/queryObserver.ts#L439-L605) (I think I link the right place here).


This is AFAIK the only way to trigger fetch of the query's data on the server, since hooks won't be called there. An alternative would be to do `suspense: true` but I don't want to go down that route yet until React 18 is stable (and then I might enable suspense as the _only_ way to query.

Reply:
In essence - unsure how I should correctly hydrate/dehydrate the query cache when doing SSR in order for errors to actually be status=error.

Reply:
this is pretty cool - I have seen a [discussion](https://github.com/tannerlinsley/react-query/discussions/3270#discussioncomment-2120752) about react-query and ssr-prepass, and I'm still hoping someone can come up with an example for the docs.

> Since the useQuery() is mounted with some (in this case, erroneous) data in the cache, it will go through the optimisticResults-flow which never really sets status='error' [as far as I can tell](https://github.com/tannerlinsley/react-query/blob/16783feed17a9361e4263edc8fecf564bc127de7/src/core/queryObserver.ts#L439-L605) (I think I link the right place here).

That I don't quite understand. If a query fails to prefetch, it should be in the cache in the correct error state. Yes, it runs through the optimistic result process, but that should only set the state optimistically to `loading` if it doesn't have data yet and will fetch once the effect that starts fetching kicks in. This just avoids that every `useQuery` call starts off in `idle` state, followed by a `loading` state 🤔 . Maybe this is an issue in RQ ?

Reply:
we'd need to verify with a test case. is it like:

- we prefetch with `prefetchQuery`, which fails
- then we mount the query with `useQuery` and would expect it to be in `error` status?

I think it would indeed go to `loading` state because the query would refetch due to `retryOnMount`, which defaults to `true`:

> retryOnMount: boolean
If set to false, the query will not be retried on mount if it contains an error. Defaults to true.

or did I get the flow wrong?

Reply:
I haven't thought of `retryOnMount` - gonna do some more experimentation.